amaranth/hdl/_time.py

```diff
@@ -0,0 +1,249 @@
+import numbers
+import re
+
+
+__all__ = ["Period"]
+
+
+class Period:
+    def __init__(self, **kwargs):
+        if not kwargs:
+            self._femtoseconds = 0
+            return
+
+        if len(kwargs) > 1:
+            raise TypeError("Period accepts at most one argument")
+
+        (unit, value), = kwargs.items()
+
+        m = re.match(r"^([munpf]?)s|([kMG]?)Hz$", unit)
+
+        if m is None:
+            raise TypeError(f"{unit} is not a valid unit")
+
+        if not isinstance(value, numbers.Real):
+            raise TypeError(f"{unit} value must be a real number")
+
+        s_unit, hz_unit = m.groups()
+
+        if s_unit is not None:
+            self._femtoseconds = round(value * {
+                "":  1_000_000_000_000_000,
+                "m": 1_000_000_000_000,
+                "u": 1_000_000_000,
+                "n": 1_000_000,
+                "p": 1_000,
+                "f": 1,
+            }[s_unit])
+
+        else:
+            if value == 0:
+                raise ZeroDivisionError("Frequency can't be zero")
+            elif value < 0:
+                raise ValueError("Frequency can't be negative")
+
+            self._femtoseconds = round(1 / value * {
```

Comment:
this is an OCD-grade nit, but `(a / b)` has less rounding error than `1 / b * a`